llpc/context/llpcCompiler.cpp

```diff
@@ -1169,29 +1208,61 @@ unsigned GraphicsShaderCacheChecker::check(const Module *module, unsigned stageM
   MetroHash::Hash fragmentHash = {};
   MetroHash::Hash nonFragmentHash = {};
   Compiler::buildShaderCacheHash(m_context, stageMask, stageHashes, &fragmentHash, &nonFragmentHash);
+  Vkgc::HashId fragmentHashId = {};
+  Vkgc::HashId nonFragmentHashId = {};
+  memcpy(&fragmentHashId.bytes, &fragmentHash.bytes, sizeof(fragmentHash));
+  memcpy(&nonFragmentHashId.bytes, &nonFragmentHash.bytes, sizeof(nonFragmentHash));
 
   IShaderCache *appCache = nullptr;
 #if LLPC_CLIENT_INTERFACE_MAJOR_VERSION < 38 || LLPC_ENABLE_SHADER_CACHE
   auto pipelineInfo = reinterpret_cast<const GraphicsPipelineBuildInfo *>(m_context->getPipelineBuildInfo());
   appCache = reinterpret_cast<IShaderCache *>(pipelineInfo->pShaderCache);
 #endif
-  if (stageMask & shaderStageToMask(ShaderStageFragment)) {
-    m_fragmentCacheEntryState = m_compiler->lookUpShaderCaches(appCache, &fragmentHash, &m_fragmentElf,
-                                                               &m_fragmentShaderCache, &m_hFragmentEntry);
-  }
+  Vkgc::ICache* pUserCache = nullptr;
+#if LLPC_CLIENT_INTERFACE_MAJOR_VERSION > 40
+  auto pipelineInfo = reinterpret_cast<const GraphicsPipelineBuildInfo *>(m_context->getPipelineBuildInfo());
+  pUserCache = pipelineInfo->pCache;
+#endif
 
-  if (stageMask & ~shaderStageToMask(ShaderStageFragment)) {
-    m_nonFragmentCacheEntryState = m_compiler->lookUpShaderCaches(appCache, &nonFragmentHash, &m_nonFragmentElf,
-                                                                  &m_nonFragmentShaderCache, &m_hNonFragmentEntry);
-  }
+  if (m_compiler->IsICacheValid())
+  {
+
+    if (stageMask & shaderStageToMask(ShaderStageFragment))
+      m_fragmentCacheResult = m_compiler->LookUpCaches(pUserCache, &fragmentHashId, &m_fragmentElf,
+                                                       &m_fragmentEntry);
 
-  if (m_nonFragmentCacheEntryState != ShaderEntryState::Compiling) {
-    // Remove non-fragment shader stages.
-    stageMask &= shaderStageToMask(ShaderStageFragment);
+    if (stageMask & ~shaderStageToMask(ShaderStageFragment))
+      m_nonFragmentCacheResult = m_compiler->LookUpCaches(pUserCache, &nonFragmentHashId, &m_nonFragmentElf,
+                                                          &m_nonFragmentEntry);
+
+    if (m_nonFragmentCacheResult != Vkgc::Result::NotFound)
+        // Remove non-fragment shader stages.
+        stageMask &= shaderStageToMask(ShaderStageFragment);
+
+    if (m_fragmentCacheResult != Vkgc::Result::NotFound)
+        // Remove fragment shader stages.
+        stageMask &= ~shaderStageToMask(ShaderStageFragment);
   }
-  if (m_fragmentCacheEntryState != ShaderEntryState::Compiling) {
-    // Remove fragment shader stages.
-    stageMask &= ~shaderStageToMask(ShaderStageFragment);
+  else
```

Comment:
The code is almost identical in both branches except the lookup method name. Can you wrap the lookup method with a lambda to not duplicate the code?

Reply:
changed.